modules/packages/LockFreeStack.chpl

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+module LockFreeStack {
+  use EpochManager;
+  use LocalAtomics;
+
+  class node {
+    type eltType;
+    var val : eltType;
+    var next : unmanaged node(eltType);
+
+    proc init(val : ?eltType) {
+      this.eltType = eltType;
+      this.val = val;
+    }
+
+    proc init(type eltType) {
+      this.eltType = eltType;
+    }
+  }
+
+  class LockFreeStack {
+    type objType;
+    var _top : LocalAtomicObject(unmanaged node(objType));
+    var _manager = new owned EpochManager();
+
+    proc init(type objType) {
+      this.objType = objType;
+    }
+
+    proc getToken() {
```

Comment:
Why would a user call this function?

Reply:
You require a token to access the data structure to pin/unpin; it's a crucial optimization. The user calls `getToken` to register their task, and this is used in pinning.

Reply:
Right, so if this a user-facing module, there needs to be chpldoc on the module talking about that and showing an example.